### PR TITLE
[PoA] consensus rules, bugfixes, refactoring, tests

### PR DESCRIPTION
**Pull request by noescape00**
Fix: #2314
Fix: #2313
Fix: #2312
Fix: #2315

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.PoA/ConsensusRules/PoACoinviewRule.cs [hunk 1]
```diff
@@ -20,7 +27,12 @@ protected override bool IsProtocolTransaction(Transaction transaction)
         /// <inheritdoc/>
         public override void CheckBlockReward(RuleContext context, Money fees, int height, Block block)
         {
-            if (block.Transactions[0].TotalOut > fees)
+            Money reward = Money.Zero;
+
+            if (height == this.network.Consensus.PremineHeight)
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.PoA/ConsensusRules/PoACoinviewRule.cs]: I think GetProofOfWorkReward should throw an exception saying there is no pow reward.

File: src/Stratis.Bitcoin.Features.PoA/SlotsManager.cs [hunk 2]
```diff
@@ -77,6 +76,7 @@ public uint GetMiningTimestamp(uint currentTime)
             return nextTimestampForMining;
         }
 
+        /// <summary>Determines whether timestamp is divisible by <see cref="PoANetwork.TargetSpacingSeconds"/>.</summary>
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.PoA/SlotsManager.cs]: ` is divisible by <see cref="PoANetwork.TargetSpacingSeconds"/>` is an implementation detail.
"Determines whether timestamp is valid according to the network rules." is better.

File: src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs [hunk 3]
```diff
@@ -158,14 +195,39 @@ private async Task CreateBlocksAsync()
...
+                catch (Exception exception)
+                {
+                    this.logger.LogCritical("Exception occurred during mining: {0}", exception.ToString());
+                    break;
+                }
             }
         }
 
+        /// <summary>Gets scriptPubKey from the wallet.</summary>
+        private Script GetPremineScriptPubKey()
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs]: Who owns this wallet?
the best way would be to have the premine sent to a multisig address owned by the members.

File: src/Stratis.Bitcoin.Features.PoA/KeyTool.cs [hunk 4]
```diff
@@ -17,33 +15,37 @@ public KeyTool(DataFolder dataFolder)
             this.dataFolder = dataFolder;
         }
 
+        /// <summary>Generates new private key.</summary>
```
[the remark this answers is not in the document]
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.PoA/KeyTool.cs]: fixed

File: src/Stratis.Bitcoin.Features.PoA/KeyTool.cs [hunk 5]
```diff
@@ -17,33 +15,37 @@ public KeyTool(DataFolder dataFolder)
...
         {
             var mnemonic = new Mnemonic(Wordlist.English, WordCount.Twelve);
-            Key privKey = mnemonic.DeriveExtKey().PrivateKey;
+            Key privateKey = mnemonic.DeriveExtKey().PrivateKey;
 
-            return privKey;
+            return privateKey;
         }
 
+        /// <summary>Gets default path for private key saving and loading.</summary>
```
[the remark this answers is not in the document]
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.PoA/KeyTool.cs]: fixed

File: src/Stratis.Bitcoin.Features.PoA/KeyTool.cs [hunk 6]
```diff
@@ -17,33 +15,37 @@ public KeyTool(DataFolder dataFolder)
...
-            stream.ReadWrite(ref privKey);
+            stream.ReadWrite(ref privateKey);
 
             ms.Seek(0, SeekOrigin.Begin);
             FileStream fileStream = File.Create(this.GetPrivateKeySavePath());
             ms.CopyTo(fileStream);
             fileStream.Close();
         }
 
+        /// <summary>Loads private key from default path.</summary>
```
[the remark this answers is not in the document]
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.PoA/KeyTool.cs]: fixed

File: src/Stratis.Bitcoin.Features.PoA/PoABlockDefinition.cs [hunk 7]
```diff
@@ -5,16 +5,14 @@
...
 using Stratis.Bitcoin.Features.Miner;
+using Stratis.Bitcoin.Features.PoA.ConsensusRules;
 using Stratis.Bitcoin.Mining;
 using Stratis.Bitcoin.Utilities;
 
 namespace Stratis.Bitcoin.Features.PoA
 {
     public class PoABlockDefinition : BlockDefinition
     {
-        /// <summary>Instance logger.</summary>
```
[the remark this answers is not in the document]
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.PoA/PoABlockDefinition.cs]: this logger wasn't used

**Pull request merged**